### PR TITLE
Add/remove networks when registry changes

### PR DESCRIPTION
**Pull request by dcbw**
Add a new -watch-networks option to watch for registry (either
remote or local) changes and add/remove networks for those events.
If this option is given along with -networks, only the networks
specified in -networks will be acted upon.  If -networks is
empty, then all network events will be acted upon.

### CODE REVIEW

File: network/manager.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+				if !m.isNetAllowed(netname) {
+					continue
+				}
+
+				switch e.Type {
+				case subnet.EventAdded:
+					if _, ok := m.networks[netname]; ok {
+						continue
+					}
+				log.Warningf("Adding network %v", netname)
```
Comment (eyakubovich) [inline network/manager.go]: gofmt

File: network/manager.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+	}
+
+	// rename(2) the temporary file to the desired location so that it becomes
+	// atomically visible with the contents
+	return os.Rename(tempFile, path)
+}
+
+func (m *Manager) RunNetwork(net *Network, wg sync.WaitGroup) {
+	wg.Add(1)
+	defer wg.Done()
```
Comment (eyakubovich) [inline network/manager.go]: Stylistic nit: so far I have been following the convention of putting the wait group's add/done calls outside:

```
go func() {
    wg.Add(1)
    defer wg.Done()
    nm.RunNetwork(net)
}
```

I'm not sure if it's "the right way" but would like to keep it for consistency.

File: network/manager.go [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+		}
+
+				log.Warningf("Running network %v", net.Name)
+		net.Run()
+		log.Infof("%v exited", net.Name)
+	}
+}
+
+func (m *Manager) watchNetworks(ctx context.Context, wg sync.WaitGroup) {
+	wg.Add(1)
```
Comment (eyakubovich) [inline network/manager.go]: Same comment as above.

File: network/manager.go (hunk 2 above)
Comment (eyakubovich) [inline network/manager.go]: I just updated the comment above to reflect that `wg.Add(1)` should be before starting the goroutine to avoid a race.

**Comment by eyakubovich** [comment]
@dcbw Apologies for slow review -- away at ContainerCon and such.

**Comment by dcbw** [comment]
> @dcbw Apologies for slow review -- away at ContainerCon and such.

No problem; I was on vacation last week anyway.

File: network/manager.go (hunk 1 above)
Comment (dcbw) [inline network/manager.go]: > gofmt. Also, make these Infos instead of Warnings?

That was actually a left-over from debugging that I've since removed.

**Comment by dcbw** [comment]
Rebased and repushed with fixes for comments...

File: network/network.go [hunk 4]
```diff
@@ -26,30 +26,36 @@ import (
...
+
+	sm         subnet.Manager
+	ipMasq     bool
+	be         backend.Backend
+	lease      *subnet.Lease
 }
 
-func New(sm subnet.Manager, name string, ipMasq bool) *Network {
+func NewNetwork(sm subnet.Manager, name string, ipMasq bool) *Network {
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (eyakubovich) [inline network/network.go]: I recently realized that a slicker way is to wrap ctx with another `WIthCancel` (https://godoc.org/golang.org/x/net/context#WithCancel). This way you get another `cancel` function but the outer one still works. This way you don't have to:

```
<-ctx.Done()
for _, net := range m.networks {
    net.Cancel()
}
```

File: network/manager.go [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+			daemon.SdNotify("READY=1")
+		}
+
+				log.Warningf("Running network %v", net.Name)
+		net.Run()
+		log.Infof("%v exited", net.Name)
+	}
+}
+
+func (m *Manager) watchNetworks(ctx context.Context, wg sync.WaitGroup) {
```
Comment (eyakubovich) [inline network/manager.go]: passing `WaitGroup` by value actually copies it and I think will cause trouble. Just create a new `WaitGroup` here and wait for it before returning from this function.

File: network/network.go (hunk 4 above)
Comment (dcbw) [inline network/network.go]: Oh nice.  Done.

**Comment by dcbw** [comment]
Rebased on backend-split-init, updated for your comments, and a further fix to the previous getNetworks() functionality to match what etcd actually returns.

**Comment by eyakubovich** [comment]
@dcbw Can you rebase again on top of master (just merged your singleton PR). This will make it easier to review.

**Comment by dcbw** [comment]
@eyakubovich will do.  I'm working on a few fixups first though that were impacting further testing of this branch.  Hope to get those and another push of this branch done today.

File: network/manager.go [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+			}
+		} else {
+			if err := writeSubnetFile(opts.subnetFile, net.Config.Network, m.ipMasq, sn); err != nil {
+				log.Warningf("%v failed to write subnet file: %s", net.Name, err)
+				return
+			}
+			daemon.SdNotify("READY=1")
+		}
+
+				log.Warningf("Running network %v", net.Name)
```
Comment (eyakubovich) [inline network/manager.go]: `log.Infof` and also fix indent / run gofmt

**Comment by eyakubovich** [comment]
@dcbw LGTM. Will merge after that indent fix.

**Comment by dcbw** [comment]
Repushed with fixups; also added network package to 'test' to get gofmt checking.  Should the 'gofmt -l' be changed to 'gofmt -d' to display diffs?  That's been useful to me in the past.

**Pull request merged**